### PR TITLE
Application Agent watches Service Bindings' resources

### PR DESCRIPTION
**Pull request by Kartikey-star**
[APPSVC-1322](https://issues.redhat.com/browse/APPSVC-1322)

### CODE REVIEW

File: test/acceptance/features/serviceBinding.feature [hunk 1]
```diff
@@ -165,3 +165,77 @@ Feature: Bind application to the secret pushed by agent app controller
                 kind: Deployment
         """
         And On Primaza Cluster "main", file "/bindings/newapp-binding/username" is unavailable in application pod running in namespace "applications"
+
+    Scenario: Service binding projection works for multiple applications matching the labels
```
Comment (filariow) [inline test/acceptance/features/serviceBinding.feature:169]: seems to me the title is not up to date. Isn't it?

File: controllers/agents/app/servicebinding_controller.go [hunk 2]
```diff
@@ -555,3 +595,132 @@ func (r *ServiceBindingReconciler) unbindApplications(ctx context.Context, req c
...
+				return
+			}
+			applicationResource := obj.(*unstructured.Unstructured)
+			if applicationResource.GetName() == constants.ApplicationAgentDeploymentName {
+				return
+			}
+			var sb primazaiov1alpha1.ServiceBinding
+			if err := r.Get(ctx, types.NamespacedName{Name: serviceBinding.Name, Namespace: serviceBinding.Namespace}, &sb); err == nil {
+				// applications are deleted, so setting the service binding status to false and reconcile
+				if errUpdateStatus := r.setStatus(ctx, sb.Spec.ServiceEndpointDefinitionSecret, sb, metav1.ConditionFalse, conditionGetAppsFailureReason, primazaiov1alpha1.ServiceBindingStateMalformed, "application deleted"); errUpdateStatus != nil {
```
Comment (filariow) [inline controllers/agents/app/servicebinding_controller.go:692]: I'd propose to break this line

File: controllers/agents/app/servicebinding_controller.go [hunk 3]
```diff
@@ -555,3 +595,132 @@ func (r *ServiceBindingReconciler) unbindApplications(ctx context.Context, req c
...
+				return
+			}
+			applicationResourceList = append(applicationResourceList, *applicationResource)
+			l.Info(fmt.Sprintf("application resource %v", applicationResourceList))
+			if psSecret, err = r.GetSecret(ctx, serviceBinding, applicationResourceList); err != nil {
+				return
+			}
+			if err := r.PrepareBinding(ctx, serviceBinding, applicationResourceList, &psSecret); err != nil {
+				return
+			}
```
Comment (filariow) [inline controllers/agents/app/servicebinding_controller.go:660]: should we retry, or log the error or update status conditions? applies also below

File: controllers/agents/app/servicebinding_controller.go [hunk 4]
```diff
@@ -87,20 +118,27 @@ func (r *ServiceBindingReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+	}
 	l.Info("Check If service binding is deleted")
 	if serviceBinding.HasDeletionTimestamp() {
 		if controllerutil.ContainsFinalizer(&serviceBinding, ServiceBindingFinalizer) {
-			applications, err := r.getApplication(ctx, req, serviceBinding, secretName)
-			if err != nil {
-				// error retrieving the application(s), so setting the service binding status to false and reconcile
-				err := r.setStatus(ctx, secretName, serviceBinding, metav1.ConditionFalse, conditionGetAppsFailureReason, primazaiov1alpha1.ServiceBindingStateMalformed, err.Error())
-				return ctrl.Result{}, err
+			// need to stop the informers if the service class is deleted
```
Comment (filariow) [inline controllers/agents/app/servicebinding_controller.go:132]: may we move this block in a `finalize` func?

File: controllers/agents/app/servicebinding_controller.go [hunk 5]
```diff
@@ -113,44 +151,61 @@ func (r *ServiceBindingReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
 		return ctrl.Result{}, err
 	}
 
+	if err := ctrl.SetControllerReference(&serviceBinding, &psSecret, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, r.Update(ctx, &psSecret)
+}
+
+func (r *ServiceBindingReconciler) GetSecret(ctx context.Context, serviceBinding v1alpha1.ServiceBinding, applications []unstructured.Unstructured) (v1.Secret, error) {
```
Comment (filariow) [inline controllers/agents/app/servicebinding_controller.go:182]: I'd propose to use `*v1.Secret` as return type. WDYT?

```suggestion
func (r *ServiceBindingReconciler) GetSecret(ctx context.Context, serviceBinding v1alpha1.ServiceBinding, applications []unstructured.Unstructured) (*v1.Secret, error) {
```

**Review by filariow — CHANGES_REQUESTED** [review]
some nits

File: test/acceptance/features/serviceBinding.feature (hunk 1 above)
Comment (Kartikey-star) [reply test/acceptance/features/serviceBinding.feature:169]: `Application Agent watches ServiceBindings' resources and projection works for application created after ServiceBinding creation` ?

File: controllers/agents/app/servicebinding_controller.go (hunk 3 above)
Comment (Kartikey-star) [reply controllers/agents/app/servicebinding_controller.go:660]: `AddEventHandler` methods is implementing interface `ResourceEventHandler ` which does not allow error to be returned.

File: controllers/agents/app/servicebinding_controller.go (hunk 5 above)
Comment (Kartikey-star) [reply controllers/agents/app/servicebinding_controller.go:182]: Makes sense

**Review by dperaza4dustbit — APPROVED** [review]
lgtm

File: test/acceptance/features/serviceBinding.feature (hunk 1 above)
Comment (filariow) [reply test/acceptance/features/serviceBinding.feature]: looks good

**Pull request merged**